pkg/controller/flinkapplication/controller.go

```diff
@@ -82,10 +83,52 @@ func (r *ReconcileFlinkApplication) getReconcileResultForError(err error) reconc
 	}
 }
 
+type SyncMap struct {
+	mx sync.RWMutex
+	m  map[string]*sync.Mutex
+}
+
+func (c *SyncMap) Get(ctx context.Context, key string) (*sync.Mutex, bool) {
+	c.mx.RLock()
+	defer c.mx.RUnlock()
+	val, found := c.m[key]
+	return val, found
+}
+
+func (c *SyncMap) GetOrLoad(ctx context.Context, key string, lock *sync.Mutex) *sync.Mutex {
+	if val, found := c.Get(ctx, key); found {
+		return val
+	} else {
+		c.mx.Lock()
+		defer c.mx.Unlock()
+		if val, found := c.m[key]; found {
+			return val
+		} else {
+			logger.Debugf(ctx, "There is no mutex for key: %v", key)
+			c.m[key] = lock
+			return lock
+		}
+	}
+}
+
+var locks = &SyncMap{
+	mx: sync.RWMutex{},
+	m:  map[string]*sync.Mutex{},
+}
+
 func (r *ReconcileFlinkApplication) Reconcile(request reconcile.Request) (reconcile.Result, error) {
 	ctx := context.Background()
 	ctx = contextutils.WithNamespace(ctx, request.Namespace)
 	ctx = contextutils.WithAppName(ctx, request.Name)
+
+	key := string(request.Namespace) + "." + string(request.Name)
```

Comment:
Please use `fmt.Sprintf`:

```suggestion
	key := fmt.Sprintf("%s.%s", request.Namespace, request.Name)
```